File: app/src/utils/initLendingPool.ts
import * as anchor from '@coral-xyz/anchor';
import { Connection, PublicKey, SystemProgram } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID, ASSOCIATED_TOKEN_PROGRAM_ID } from '@solana/spl-token';
import fallIdl from '../idl/fall.json';
import { Idl } from '@coral-xyz/anchor';
import { web3 } from '@project-serum/anchor';

import {
  LENDING_SEED,
  LENDING_AUTHORITY_SEED,
  LENDING_TOKEN_SEED,
  BORROW_TOKEN_SEED,
  COLLATERAL_TOKEN_SEED,
  LENDING_HEIGHT_TOKEN_SEED,
  BORROW_HEIGHT_TOKEN_SEED,
} from '../utils/constants';

export async function initLendingPool(
  wallet: any,
  connection: Connection,
  poolPda: PublicKey,
) {
  try {
    console.log('Creating lending pool...');
    const provider = new anchor.AnchorProvider(
      connection,
      wallet,
      { preflightCommitment: "confirmed" }
    );

    const program = new anchor.Program(
      fallIdl as Idl,
      provider
    );

    console.log('Fetching pool info...');
    const pool = await program.account.pool.fetch(poolPda);
    console.log('Pool info:', {
      amm: pool.amm.toString(),
      mintA: pool.mintA.toString(),
      mintB: pool.mintB.toString(),
    });

    const mintA = pool.mintA;
    const mintB = pool.mintB;

    // 获取 Lending Pool PDA
    const [lendingPoolPda] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), mintA.toBuffer(), mintB.toBuffer(), Buffer.from(LENDING_SEED)],
      program.programId
    );

    // 获取 Lending Pool Authority PDA
    const [lendingPoolAuthorityPda] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), mintA.toBuffer(), mintB.toBuffer(), Buffer.from(LENDING_AUTHORITY_SEED)],
      program.programId
    );

    // 获取 Lending Receipt Token Mint PDA
    const [lendingReceiptTokenMint] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(),  Buffer.from(LENDING_TOKEN_SEED)],
      program.programId
    );

    // 获取 Borrow Receipt Token Mint PDA
    const [borrowReceiptTokenMint] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), Buffer.from(BORROW_TOKEN_SEED)],
      program.programId
    );

    // 获取 Collateral Receipt Token Mint PDA
    const [collateralReceiptTokenMint] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), Buffer.from(COLLATERAL_TOKEN_SEED)],
      program.programId
    );

    // 获取 Lending Pool Token Accounts
    const lendingPoolAccountA = await anchor.utils.token.associatedAddress({
      mint: mintA,
      owner: lendingPoolAuthorityPda
    });

    const lendingPoolAccountB = await anchor.utils.token.associatedAddress({
      mint: mintB,
      owner: lendingPoolAuthorityPda
    });

    // Block Height Token Mints
    const [lenderLendingBlockHeightTokenMint] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), Buffer.from(LENDING_HEIGHT_TOKEN_SEED)],
      program.programId
    );

    const [borrowerBorrowBlockHeightTokenMint] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), Buffer.from(BORROW_HEIGHT_TOKEN_SEED)],
      program.programId
    );

    console.log('Sending transaction...');
    const tx = await program.methods
      .initLendingPool()
      .accounts({
        pool: poolPda,
        mintA: mintA,
        mintB: mintB,
        lendingPool: lendingPoolPda,
        lendingPoolAuthority: lendingPoolAuthorityPda,
        lendingPoolTokenA: lendingPoolAccountA,
        lendingPoolTokenB: lendingPoolAccountB,
        lendingReceiptTokenMint: lendingReceiptTokenMint,
        borrowReceiptTokenMint: borrowReceiptTokenMint,
        collateralReceiptTokenMint: collateralReceiptTokenMint,
        lenderLendingBlockHeightTokenMint: lenderLendingBlockHeightTokenMint,
        borrowerBorrowBlockHeightTokenMint: borrowerBorrowBlockHeightTokenMint,
        payer: provider.wallet.publicKey,
        tokenProgram: TOKEN_PROGRAM_ID,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
        rent: web3.SYSVAR_RENT_PUBKEY,
      }).rpc();

    // 获取交易详情
    const txInfo = await connection.getTransaction(tx, {
      commitment: 'confirmed',
      maxSupportedTransactionVersion: 0
    });

    if (txInfo?.meta?.logMessages) {
      console.log("Program Logs:");
      txInfo.meta.logMessages.forEach(log => {
        console.log(log);
      });
    }

    return {
      tx,
      lendingPoolPda,
      lendingPoolAuthorityPda,
      lendingReceiptTokenMint,
      borrowReceiptTokenMint,
      collateralReceiptTokenMint,
      lendingPoolAccountA,
      lendingPoolAccountB
    };
  } catch (error) {
    console.error('Error in createLendingPool:', error);
    throw error;
  }
}

File: programs/fall/src/instructions/init_lending_pool.rs
use anchor_lang::prelude::*;
use anchor_spl::{
    associated_token::AssociatedToken,
    token::{Mint, Token},
};
use crate::{
    constants::{LENDING_TOKEN_SEED,BORROW_TOKEN_SEED,LENDING_AUTHORITY_SEED,COLLATERAL_TOKEN_SEED,LENDER_LENDING_BLOCK_HEIGHT_TOKEN_SEED,BORROWER_BORROW_BLOCK_HEIGHT_TOKEN_SEED},
    state::Pool,
};

#[derive(Accounts)]
pub struct InitLendingPool<'info> {
    /// CHECK: Read only authority
    #[account(
        seeds = [
            pool.amm.as_ref(),
            pool.mint_a.key().as_ref(),
            pool.mint_b.key().as_ref(),
        ],
        bump,
    )]
    pub pool: Box<Account<'info, Pool>>,

    /// CHECK: Read only authority
    #[account(
        seeds = [
            pool.key().as_ref(),
            LENDING_AUTHORITY_SEED,
        ],
        bump,
    )]
    pub lending_pool_authority: AccountInfo<'info>,

    #[account(
        init_if_needed,
        payer = payer,
        seeds = [
            pool.key().as_ref(),
            LENDING_TOKEN_SEED,
        ],
        bump,
        mint::decimals = 6,
        mint::authority = lending_pool_authority,
        mint::freeze_authority = lending_pool_authority,
    )]
    pub lending_receipt_token_mint: Box<Account<'info, Mint>>,

    #[account(
        init,
        payer = payer,
        seeds = [
            pool.key().as_ref(),
            BORROW_TOKEN_SEED,
        ],
        bump,
        mint::decimals = 6,
        mint::authority = lending_pool_authority,
        mint::freeze_authority = lending_pool_authority,
    )]
    pub borrow_receipt_token_mint: Box<Account<'info, Mint>>,

    #[account(
        init,
        payer = payer,
        seeds = [
            pool.key().as_ref(),
            COLLATERAL_TOKEN_SEED,
        ],
        bump,
        mint::decimals = 6,
        mint::authority = lending_pool_authority,
        mint::freeze_authority = lending_pool_authority,
    )]
    pub collateral_receipt_token_mint: Box<Account<'info, Mint>>,

    #[account(
        init,
        payer = payer,
        seeds = [
            pool.key().as_ref(),
            LENDER_LENDING_BLOCK_HEIGHT_TOKEN_SEED,
        ],
        bump,
        mint::decimals = 6,
        mint::authority = lending_pool_authority,
        mint::freeze_authority = lending_pool_authority,
    )]
    pub lender_lending_block_height_mint: Box<Account<'info, Mint>>,

    #[account(
        init,
        payer = payer,
        seeds = [
            pool.key().as_ref(),
            BORROWER_BORROW_BLOCK_HEIGHT_TOKEN_SEED,
        ],
        bump,
        mint::decimals = 6,
        mint::authority = lending_pool_authority,
        mint::freeze_authority = lending_pool_authority,
    )]
    pub borrower_borrow_block_height_mint: Box<Account<'info, Mint>>,

    #[account(mut)]
    pub payer: Signer<'info>,

    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
}

pub fn init_lending_pool(_ctx: Context<InitLendingPool>) -> Result<()> {
    Ok(())
}